src/backend/columnar/columnar_tableam.c

```diff
@@ -2126,9 +2128,87 @@ ColumnarProcessUtility(PlannedStmt *pstmt,
 
 		RelationClose(rel);
 	}
+	else if (IsA(parsetree, CreateStmt))
+	{
+		CreateStmt *createStmt = castNode(CreateStmt, parsetree);
+		if (createStmt->accessMethod != NULL &&
+			!strcmp(createStmt->accessMethod, COLUMNAR_AM_NAME))
+		{
+			columnarRangeVar = createStmt->relation;
+			createStmt->options = ExtractColumnarRelOptions(createStmt->options,
+															&columnarOptions);
+		}
+	}
+	else if (IsA(parsetree, CreateTableAsStmt))
+	{
+		CreateTableAsStmt *createTableAsStmt = castNode(CreateTableAsStmt, parsetree);
+		IntoClause *into = createTableAsStmt->into;
+		if (into->accessMethod != NULL &&
+			!strcmp(into->accessMethod, COLUMNAR_AM_NAME))
+		{
+			columnarRangeVar = into->rel;
+			into->options = ExtractColumnarRelOptions(into->options,
+													  &columnarOptions);
+		}
+	}
+	else if (IsA(parsetree, AlterTableStmt))
+	{
+		AlterTableStmt *alterTableStmt = castNode(AlterTableStmt, parsetree);
+		bool toColumnar = false;
+
+		Relation rel = relation_openrv_extended(alterTableStmt->relation, AccessShareLock,
+												alterTableStmt->missing_ok);
+
+		if (rel != NULL){
+		bool isColumnar = rel->rd_tableam == GetColumnarTableAmRoutine();
+
+		ListCell *lc;
+		foreach (lc, alterTableStmt->cmds)
+		{
+			AlterTableCmd *alterTableCmd = castNode(AlterTableCmd, lfirst(lc));
+
+			if (alterTableCmd->subtype == AT_SetRelOptions ||
+				alterTableCmd->subtype == AT_ResetRelOptions)
+			{
+				List *options = castNode(List, alterTableCmd->def);
+
+				if (isColumnar || toColumnar)
+				{
+					columnarRangeVar = alterTableStmt->relation;
+					alterTableCmd->def = (Node *) ExtractColumnarRelOptions(options,
+																			&columnarOptions);
```

Comment:
We are looping over `AlterTableCmd`'s. With the current implementation of `ExtractColumnarRelOptions` we overwrite the `columnarOptions` variable instead of appending the options to the list.

Could there be a posibility of hitting the `ExtractColumnarRelOptions` call here more then once? In that case we will _only_ have the last options extracted.

Alternatively we could change the behaviour of `ExtractColumnarRelOptions` to _not_ swap the lists, but instead `lappend` directly to the out parameter.

If this could only ever be hit once, would it make sense to break out of the loop? Guess that is not a possibility due to the PG15 block below. Then an `elog` when we hit this code wile `columnarOptions` is not empty anymore would help guard against undesired behaviour.

Reply:
Fixed (and tested). New settings are appended to the list, and the last update wins.